Add contribution guideline (#166)
matterpoll/matterpoll: CONTRIBUTING.md

## Reporting Issues

If you think you found a bug, [please use the GitHub issue tracker](https://github.com/matterpoll/matterpoll/issues/new?template=bug_report.md) to open an issue. When opening an issue, please provide the required information in the issue template.

## Translating strings

Matterpoll supports localization to various languages. We as maintainers rely on contributors to help with the translations.

The localization process is defined below:
- During development, new translation strings may be added or existing ones updated.
- When a new version is planned to release soon, a maintainer opens an issue informing about this. The maintainer will ping all translation maintainer to inform them about this.
- Translation maintainers submit PRs with new translations, which may get reviewed by other translators.
- After all translation PRs are merged, the new version is released.

### Translation Maintainers

- France: [@ldidry](https://github.com/ldidry)
- German: [@hanzei](https://github.com/hanzei)
- Japanese: [@kaakaa](https://github.com/kaakaa/)

### Translation Process for Existing Languages

> 1. Ensure all translation messages are correctly extracted:

Reviewer: It's better to add the process of installing `go-i18n`, or add new section about requirements for translating.

Author: Added a paragraph about this.

`goi18n extract -format json -outdir assets/i18n/ server/`

2. Update your translation files:

`goi18n merge -format json -outdir assets/i18n/ assets/i18n/active.*.json`

3. Translate all messages in `asserts/i18n/translate.*.json` for the languages you are comfortable with.

4. Merge the translated messages into the active message files:

`goi18n merge -format json -outdir assets/i18n/ assets/i18n/active.*.json assets/i18n/translate.*.json`

5. Commit **only the language files you touched** and [submit a PR](https://github.com/matterpoll/matterpoll/compare).

### Translation Process for New Languages

Let's say you want to translate the local `de`. Replace  `de` in the following commands with the local you want to translate. See [here](https://github.com/mattermost/mattermost-server/tree/master/i18n) for the list of possible locals.

> 1. Create a translation file:

Reviewer: Same is here.

# Contributing to Matterpoll

Thank you for your interest in contributing! Join the join the [**Matterpoll channel**](https://community.mattermost.com/core/channels/matterpoll) on the Mattermost community server for discussion about this plugin.

## Reporting Issues

If you think you found a bug in Hugo, [please use the GitHub issue tracker](https://github.com/matterpoll/matterpoll/issues/new?template=bug_report.md) to open an issue. When opening an issue, please provide the required information in the issue template.

## Helping with translations

Matterpoll supports localization in various languages. Because we as the maintainer only speak a small amount of language, we rely on contributors to help us with the translations.

The translations process is:
- While developing, new translation messages may be added or existing ones changed.

> - When a new version will releases soon, a maintainer will open an issue informing about this.

Reviewer 2: ```suggestion
- When a new version is planned to release soon, a maintainer opens an issue informing about this.
```

> - When a new version will releases soon, a maintainer will open an issue informing about this.

Reviewer 2: And I assume this is the repo maintainer, not translation maintainer?

> - When a new version will releases soon, a maintainer will open an issue informing about this.

Author: > Is there an example issue?

No. Currently the process is just defined, but we havn't tried it out.

> And I assume this is the repo maintainer, not translation maintainer?

Correct. :+1: I updated the text to make this more clear.

> - The maintainer will ping all translation maintainer to inform them about this.

Author: Made this more clear.

> - They open PR's with new translations, which (may) get reviewed by other translators.

Reviewer 2: ```suggestion
- Translation maintainers submit PRs with new translations, which may get reviewed by other translators.
```

> - They open PR's with new translations, which (may) get reviewed by other translators.

Author: Currently not.

> - After all translation PR's are merged, the new version is released.

Reviewer 2: ```suggestion
- After all translation PRs are merged, the new version is released.
```

> - After all translation PR's are merged, the new version is released.

Author: Good question.

We have no hard deadline when we want to release a version. To we can wait "a bit". But if no PR is submitted, we release without it.

> - After all translation PR's are merged, the new version is released.

Reviewer: I totally agree with @hanzei. If some messages are not translated, it's just be displayed in English and it's not critical bug.

And I'm wondering that `a bit` is about a week.

> - After all translation PR's are merged, the new version is released.

Author: Update the text to reflect this

Translate all messages in `asserts/i18n/translate.*.json` for the languages you are comfortable with.

Merge the translated messages into the active message files:

`goi18n merge -format json -outdir assets/i18n/ assets/i18n/active.*.json assets/i18n/translate.*.json`

Commit **only the language files you touched** and [submit a PR](https://github.com/matterpoll/matterpoll/compare).

### Translating a new language

Let's say you want to translate the local `de`. Replace  `de` in the following commands with the local you want to translate. See [here](https://github.com/mattermost/mattermost-server/tree/master/i18n) for the list of possible locals.

Create a translation file:

`touch asserts/i18n/translate.de.json`

Merge all current messages into your translation file:

`goi18n merge -format json -outdir assets/i18n/ assets/i18n/active.en.json assets/i18n/translate.de.json`

Translate all messages in `asserts/i18n/translate.de.json` and rename it to `active.de.json`.

[Submit a PR](https://github.com/matterpoll/matterpoll/compare) with this file and add you to the list of [Translation Maintainer](#translation-maintainer)

## Submitting Patches

> If you contributing a feature, [please open a feature request](https://github.com/matterpoll/matterpoll/issues/new?template=feature_request.md) first. This way the feature can be discussed and fully specified before you start working on this. Small code changes can be submitted without opening an issue first.

Author: > I wonder what defines as a "small" issue. We could use a guideline referred to here developers.mattermost.com/contribute/getting-started/contributions-without-ticket

I see your point. But for now this is fine. There hasn't been a lot of contributions jet. We can improve the definition if this becomes a problem.

## Reporting Issues

If you think you found a bug, [please use the GitHub issue tracker](https://github.com/matterpoll/matterpoll/issues/new?template=bug_report.md) to open an issue. When opening an issue, please provide the required information in the issue template.

## Translating strings

Matterpoll supports localization to various languages. We as maintainers rely on contributors to help with the translations.

Matterpoll uses [go-i18n](https://github.com/nicksnyder/go-i18n) as library and tool to manage translation. The CLI tool `goi18n` is required to manage translation. You can install it by running `env GO111MODULE=off go get -u github.com/nicksnyder/go-i18n/v2/goi18n`.

The localization process is defined below:
- During development, new translation strings may be added or existing ones updated.
- When a new version is planned to release soon, a repository maintainer opens an issue informing about this. The maintainer will ping all translation maintainer to inform them about this.
- Translation maintainers submit PRs with new translations, which may get reviewed by other translators.

> - After all translation PRs are merged, the new version is released.

Reviewer 2: @hanzei It was about the translations PRs and how long we'd wait for translation maintainers to submit them

Author: Please see my answer at https://github.com/matterpoll/matterpoll/pull/166#discussion_r295316022

> - After all translation PRs are merged, the new version is released. If a translation PR is not submitted within in week, the release will be cut anyway.

Reviewer 2: ```suggestion
- After all translation PRs are merged, the new version is released. If a translation PR is not submitted within a week, the release is cut without it.
```

Reviewer 2: Just minor proposal, otherwise LGTM!
